### PR TITLE
Feature/f acc

### PR DESCRIPTION
**Pull request by reguly**
Bring in OpenACC support in Fortran, using with the CUDA backend.

### CODE REVIEW

File: translator/fortran/python/op2_gen_openacc.py [hunk 1]
```diff
@@ -0,0 +1,1038 @@
...
+          for d in range(0,int(dims[g_m])):
+            if accs[g_m] == OP_INC:
+              code('opDat'+str(g_m+1)+'Local_'+str(d+1)+' = opDat'+str(g_m+1)+'Local_'+str(d+1)+' + opDat'+str(g_m+1)+'LocalArr('+str(d+1)+')')
+            if accs[g_m] == OP_MIN:
+              code('opDat'+str(g_m+1)+'Local_'+str(d+1)+' = MIN(opDat'+str(g_m+1)+'Local_'+str(d+1)+', opDat'+str(g_m+1)+'LocalArr('+str(d+1)+'))')
+            if accs[g_m] == OP_MAX:
+              code('opDat'+str(g_m+1)+'Local_'+str(d+1)+' = MAX(opDat'+str(g_m+1)+'Local_'+str(d+1)+', opDat'+str(g_m+1)+'LocalArr('+str(d+1)+'))')
+
+    for g_m in range(0,nargs):
+        if stage_flags[g_m] == 1 and accs[g_m] <> OP_READ:
```
Comment (kynan) [inline translator/fortran/python/op2_gen_openacc.py]: Use `!=`

File: translator/fortran/python/op2_gen_openacc.py [hunk 2]
```diff
@@ -0,0 +1,1038 @@
...
+          if optflags[g_m]==1:
+            ENDIF()
+    depth = depth + 2
+
+    depth = depth - 2
+    ENDDO()
+    if not host_exec:
+      code('')
+      for g_m in range(0,nargs):
+        if maps[g_m] == OP_GBL and accs[g_m] <> OP_READ and accs[g_m] <> OP_WRITE and (not dims[g_m].isdigit() or int(dims[g_m])>1):
```
Comment (kynan) [inline translator/fortran/python/op2_gen_openacc.py]: Use `!=`

File: translator/fortran/python/op2_gen_openacc.py [hunk 3]
```diff
@@ -0,0 +1,1038 @@
...
+        code('& opDat'+str(invinds[g_m]+1)+'Local, &')
+      for g_m in range(0,nargs):
+        if maps[g_m] == OP_ID:
+          if g_m in needDimList:
+            code('& opArg'+str(g_m+1)+'%dim, &')
+          code('& opDat'+str(g_m+1)+'Local, &')
+        elif maps[g_m] == OP_GBL:
+          if g_m in needDimList:
+            code('& opArg'+str(g_m+1)+'%dim, &')
+          if accs[g_m] <> OP_READ and accs[g_m] <> OP_WRITE and not host_exec:
```
Comment (kynan) [inline translator/fortran/python/op2_gen_openacc.py]: Use `!=`

File: translator/fortran/python/op2_gen_openacc.py [hunk 4]
```diff
@@ -0,0 +1,1038 @@
...
+          if maps[g_m] == OP_MAP and (not mapnames[g_m] in k):
+            k = k + [mapnames[g_m]]
+            code('& opDat'+str(invinds[inds[g_m]-1]+1)+'Map, &')
+            code('& opDat'+str(invinds[inds[g_m]-1]+1)+'MapDim, &')
+      code('& col_reord_'+name+', exec_size, offset_b, nelem )')
+
+      if reduct and not host_exec:
+        IF('i1 .EQ. actualPlan_'+name+'%ncolors_owned -1')
+        for g_m in range(0,nargs):
+          if maps[g_m] == OP_GBL and accs[g_m] <> OP_READ and accs[g_m] <> OP_WRITE:
```
Comment (kynan) [inline translator/fortran/python/op2_gen_openacc.py]: Use `!=`

File: translator/fortran/python/op2_gen_openacc.py [hunk 5]
```diff
@@ -0,0 +1,1038 @@
...
+        if invinds[g_m] in needDimList:
+          code('& opArg'+str(invinds[g_m]+1)+'%dim, &')
+        code('& opDat'+str(invinds[g_m]+1)+'Local, &')
+      for g_m in range(0,nargs):
+        if g_m in needDimList:
+          code('& opArg'+str(g_m+1)+'%dim, &')
+        if maps[g_m] == OP_ID:
+          code('& opDat'+str(g_m+1)+'Local, &')
+        elif maps[g_m] == OP_GBL:
+          if accs[g_m] <> OP_READ and accs[g_m] <> OP_WRITE and not host_exec:
```
Comment (kynan) [inline translator/fortran/python/op2_gen_openacc.py]: Use `!=`

File: translator/fortran/python/op2_gen_openacc.py [hunk 6]
```diff
@@ -0,0 +1,1038 @@
...
+    IF('(n_upper .EQ. 0) .OR. (n_upper .EQ. opSetCore%core_size)')
+    if host_exec:
+      code('CALL op_mpi_wait_all(numberOfOpDats,opArgArray)')
+    else:
+      code('CALL op_mpi_wait_all_cuda(numberOfOpDats,opArgArray)')
+    ENDIF()
+
+    if ninds==0:
+      for g_m in range(0,nargs):
+        if maps[g_m] == OP_GBL and accs[g_m] <> OP_READ and accs[g_m] <> OP_WRITE and not host_exec:
```
Comment (kynan) [inline translator/fortran/python/op2_gen_openacc.py]: Use `!=`

File: translator/fortran/python/op2_gen_openacc.py [hunk 7]
```diff
@@ -0,0 +1,1038 @@
...
+    code('')
+    if host_exec:
+      code('CALL op_mpi_set_dirtybit(numberOfOpDats,opArgArray)')
+    else:
+      code('CALL op_mpi_set_dirtybit_cuda(numberOfOpDats,opArgArray)')
+    code('')
+
+    #reductions
+    for g_m in range(0,nargs):
+      if maps[g_m] == OP_GBL and accs[g_m] <> OP_READ and accs[g_m] <> OP_WRITE and not host_exec:
```
Comment (kynan) [inline translator/fortran/python/op2_gen_openacc.py]: Use `!=`

File: translator/fortran/python/util.py [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+          return loc2
+      loc2 = loc2 + 1
+
+const_list = []
+
+def replace_consts(text):
+  global const_list
+  i = text.find('use HYDRA_CONST_MODULE')
+  if i > -1:
+    fi2 = open("hydra_constants_list.txt","r")
```
Comment (kynan) [inline translator/fortran/python/util.py]: Use a context manager

File: translator/fortran/python/util.py [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+    else:
+      idx = mapnames.index(mapnames[g_m])
+      return 'opDat'+str(idx+1)+'_stride_OP2CONSTANT'
+
+
+def replace_soa(text,nargs,soaflags,name,maps,accs,set_name,mapnames,repl_inc,hydra,bookleaf):
+  OP_ID   = 1;  OP_GBL   = 2;  OP_MAP = 3;
+
+  OP_READ = 1;  OP_WRITE = 2;  OP_RW  = 3;
+  OP_INC  = 4;  OP_MAX   = 5;  OP_MIN = 6;
```
Comment (kynan) [inline translator/fortran/python/util.py]: Shouldn't these be defined in a central place?

File: translator/fortran/python/util.py [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+  #
+  # Apply SoA to variable accesses
+  #
+  j = text.find(name+'_gpu')
+  endj = arg_parse(text,j)
+  while text[j] <> '(':
```
Comment (kynan) [inline translator/fortran/python/util.py]: Use `!=`

File: translator/fortran/python/util.py [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
+    i = search_offset + res.start() + 4
+    #find name: whatever is in front of opening bracket
+    openbracket = i+text[i:].find('(')
+    fun_name = text[i:openbracket].strip()
+    # #if hyd_dump, comment it out
+    # if fun_name == hyd_dump:
+    #   text = text[0:search_offset + res.start()]+'!'+text[search_offset + res.start():]
+    #   search_offset = i
+    #   res=re.search('\\bcall\\b',text[search_offset:])
+    #   continue
```
Comment (kynan) [inline translator/fortran/python/util.py]: Do not add commented code

File: translator/fortran/python/util.py [hunk 12]
```diff
@@ -0,0 +1,250 @@
...
+
+    if fun_name.lower() in funlist:
+      search_offset = i
+      res=re.search('\\bcall\\b',text[search_offset:])
+      continue
+
+    #print fun_name
```
Comment (kynan) [inline translator/fortran/python/util.py]: Do not add commented code

File: translator/fortran/python/util.py [hunk 13]
```diff
@@ -0,0 +1,250 @@
...
+      curr_pos = curr_pos+text[curr_pos:].find('\n')+1
+    curr_pos = curr_pos-1
+    arglist = text[openbracket:curr_pos]
+    #find the file containing the implementation
+    subr_file =  os.popen('grep -Rilw --include "*.F95" --exclude "*kernel.*" "subroutine '+fun_name+'" . | head -n1').read().strip()
+    if (len(subr_file) == 0) or (not os.path.exists(subr_file)):
+      print 'Error, subroutine '+fun_name+' implementation not found in files, check parser!'
+      exit(1)
+    #read the file and find the implementation
+    subr_fileh = open(subr_file,'r')
```
Comment (kynan) [inline translator/fortran/python/util.py]: Use a context manager

File: translator/fortran/python/util.py [hunk 14]
```diff
@@ -0,0 +1,250 @@
...
+      writes_offset = writes_offset + writes.start()
+      subr_text = subr_text[0:writes_offset]+'!'+subr_text[writes_offset:]
+      writes_offset = writes_offset + subr_text[writes_offset:].find('\n')+1
+      while (subr_text[writes_offset:].strip()[0] == '&'):
+        subr_text = subr_text[0:writes_offset]+'!'+subr_text[writes_offset:]
+        writes_offset = writes_offset + subr_text[writes_offset:].find('\n')+1
+      writes = re.search('\\bwrite\\b',subr_text[writes_offset:])
+
+    subr_text = replace_npdes(subr_text)
+    if attr <> '':
```
Comment (kynan) [inline translator/fortran/python/util.py]: Use `!=`

**Comment by gihanmudalige** [comment]
I am getting an error compiling Fortran airfoil_hdf5 :
input.F90(155): error #5082: Syntax error, found IDENTIFIER 'ENTER' when expecting one of: DATA UPDATE TEAMS MAP DEVICE IF <END-OF-STATEMENT> ;
!$omp target enter data map(to:gam, gm1, cfl, eps, mach, alpha, qinf(4))
-------------^
compilation aborted for input.F90 (code 1)
make: *** [airfoil_hdf5_openmp] Error 1

**Comment by reguly** [comment]
Thank you, should be fixed now.

Cheers,
Istvan

**Comment by gihanmudalige** [comment]
Thanks Istvan. Seems to have fixed that, but a different error for the openacc version:

PGF90-S-0155-Procedures called in a compute region must have acc routine information: adt_calc_gpu (adt_calc_acckernel.F90: 92)
PGF90-S-0155-Accelerator region ignored; see -Minfo messages  (adt_calc_acckernel.F90: 75)
op_wrap_adt_calc:
     75, Accelerator region ignored
     92, Accelerator restriction: call to 'adt_calc_gpu' with no acc routine information
  0 inform,   0 warnings,   2 severes, 0 fatal for op_wrap_adt_calc
airfoil_hdf5_op.F90:
PGF90-F-0004-Unable to open MODULE file op2_fortran_hdf5_declarations.mod (airfoil_hdf5_op.F90: 6)
PGF90/x86-64 Linux 15.10-0: compilation aborted
make: *** [airfoil_hdf5_openacc] Error 2

**Comment by reguly** [comment]
Ah, so many backends, apologies. Pushed the update.

Cheers,
Istvan

**Comment by gihanmudalige** [comment]
Ok. this has now passed and can be merged

**Pull request merged**